### PR TITLE
feat(autoware_pointcloud_preprocessor): expansion for single lidar system

### PR DESCRIPTION
**Pull request by shtokuda**
## Description
In this PR, I added the callback function to pointcloud_preprocessor for single lidar system.
https://github.com/orgs/autowarefoundation/discussions/4700
## Tests performed
I have checked that the /sensing/lidar/concatenated/pointcloud topic is published even if the number of input_topic is 1 on the LSIM.
## Interface changes
None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_nodelet.hpp [hunk 1]
```diff
@@ -174,6 +175,9 @@ class PointCloudConcatenateDataSynchronizerComponent : public rclcpp::Node
   void cloud_callback(
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_nodelet.hpp:175]: As ```cloud_callback``` and ```single_cloud_callback``` have almost the same name. So perhaps renaming ```cloud_callback``` to ```multi_cloud_callback``` makes it clear.

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp [hunk 2]
```diff
@@ -94,12 +94,18 @@ PointCloudConcatenateDataSynchronizerComponent::PointCloudConcatenateDataSynchro
...
       return;
     }
-    if (input_topics_.size() == 1) {
-      RCLCPP_ERROR(get_logger(), "Only one topic given. Need at least two topics to continue.");
+
+    // Check if there are duplicated input topic
+    std::sort(input_topics_.begin(), input_topics_.end());
+    auto duplicate_it = std::adjacent_find(input_topics_.begin(), input_topics_.end());
+    if (duplicate_it != input_topics_.end()) {
+      RCLCPP_ERROR(get_logger(), "Each input topic must have a defferent name!");
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp:108]: Let's print the duplicated topic here as follows (also fix a typo d"e"fferent).
```
"Duplicated topic name: <topic_name>. Each input topic must have a different name."
```

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp [hunk 3]
```diff
@@ -94,12 +94,18 @@ PointCloudConcatenateDataSynchronizerComponent::PointCloudConcatenateDataSynchro
     }
     declare_parameter("input_topics", std::vector<std::string>());
     input_topics_ = get_parameter("input_topics").as_string_array();
+
+    // Check if input topics are empty
     if (input_topics_.empty()) {
-      RCLCPP_ERROR(get_logger(), "Need a 'input_topics' parameter to be set before continuing!");
+      RCLCPP_ERROR(get_logger(), "Need 'input_topics' parameter to be set before continuing!");
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp:100]: Now:
```
"Need 'input_topics' parameter to be set before continuing!"
```

My proposal:
```
"Missing 'input_topics' parameter. Skipping PointCloudConcatenateDataSynchronizerComponent initialization."
```

Honestly I'm not understanding the behavior of the ```get_logger()```. If this logger outputs the component name as default, then my proposal becomes simply ```"Missing 'input_topics' parameter"```

File: sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_nodelet.hpp (hunk 1 above)
Comment (sasakisasaki) [reply sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_nodelet.hpp:175]: (If you apply the fix, please do also for all places on where the name is used)

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp [hunk 4]
```diff
@@ -473,6 +487,64 @@ void PointCloudConcatenateDataSynchronizerComponent::publish()
   }
 }
 
+void PointCloudConcatenateDataSynchronizerComponent::publishSingleLidar()
+{
+  stop_watch_ptr_->toc("processing_time", true);
+  sensor_msgs::msg::PointCloud2::SharedPtr transformed_cloud_ptr(
+    new sensor_msgs::msg::PointCloud2());
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp:494]: Please consider to use smart pointer :+1:

**Comment by sasakisasaki** [comment]
@shtokuda As we had a discussion, perhaps the [time synchronize side](https://github.com/shtokuda/autoware.universe/blob/4f32bc1a89e2be59eec2e04796adcef6674350c6/sensing/autoware_pointcloud_preprocessor/src/time_synchronizer/time_synchronizer_nodelet.cpp#L73-L76) is also needed to be fixed.

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp [hunk 5]
```diff
@@ -162,18 +168,26 @@ PointCloudConcatenateDataSynchronizerComponent::PointCloudConcatenateDataSynchro
...
     for (size_t d = 0; d < input_topics_.size(); ++d) {
       cloud_stdmap_.insert(std::make_pair(input_topics_[d], nullptr));
-      cloud_stdmap_tmp_ = cloud_stdmap_;
 
       // CAN'T use auto type here.
-      std::function<void(const sensor_msgs::msg::PointCloud2::ConstSharedPtr msg)> cb = std::bind(
-        &PointCloudConcatenateDataSynchronizerComponent::cloud_callback, this,
-        std::placeholders::_1, input_topics_[d]);
+      std::function<void(const sensor_msgs::msg::PointCloud2::ConstSharedPtr msg)> cb;
+      if (input_topics_.size() == 1) {
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp:174]: Adding the following comment just before the if statement looks good.
```
// Single lidar case
```

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp [hunk 6]
```diff
@@ -118,17 +122,25 @@ PointCloudConcatenationComponent::PointCloudConcatenationComponent(
...
+          &PointCloudConcatenationComponent::cloud_callback, this, std::placeholders::_1,
+          input_topics_[d]);
+      }
 
       filters_[d].reset();
       filters_[d] = this->create_subscription<sensor_msgs::msg::PointCloud2>(
         input_topics_[d], rclcpp::SensorDataQoS().keep_last(maximum_queue_size_), cb);
     }
+
+    cloud_stdmap_tmp_ = cloud_stdmap_;
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp:143]: Is this ```cloud_stdmap_tmp_``` used? If not, let's remove :+1:

**Comment by sasakisasaki** [comment]
We tested the feature provided by this PR with @shtokuda and found it worked with minor fix in the time synchronize side. The video is as follows and we can see it is working only with a front single lidar.
[Screencast.webm](https://github.com/user-attachments/assets/cb684cc2-9cfa-4b66-98d1-134829b44aef)

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp [hunk 7]
```diff
@@ -446,12 +503,35 @@ void PointCloudConcatenationComponent::cloud_callback(
...
 void PointCloudConcatenationComponent::timer_callback()
 {
   using std::chrono_literals::operator""ms;
   timer_->cancel();
   if (mutex_.try_lock()) {
-    publish();
+    if (input_topics_.size() == 1) {
+      publishSingleLidar();
+    } else {
+      publish();
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp:533]: Perhaps, this ```publish()``` is triggered even when ```input_topics_.size() == 0```. Let's ensure non-zero case by:
```
} else if (input_topics_.size() > 1) {
```

not:

```
} else {
```

:+1:

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp [hunk 8]
```diff
@@ -446,12 +503,35 @@ void PointCloudConcatenationComponent::cloud_callback(
   }
 }
 
+void PointCloudConcatenationComponent::single_cloud_callback(
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & input_ptr, const std::string & topic_name)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp:509]: If possible, let's use ```mutex_.try_lock()``` for the solid error handling :+1: .

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp [hunk 9]
```diff
@@ -473,6 +487,64 @@ void PointCloudConcatenateDataSynchronizerComponent::publish()
...
+        const auto pipeline_latency_ms =
+          std::chrono::duration<double, std::milli>(
+            std::chrono::nanoseconds(
+              (this->get_clock()->now() - e.second->header.stamp).nanoseconds()))
+            .count();
+        debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+          "debug" + e.first + "/pipeline_latency_ms", pipeline_latency_ms);
+      }
+    }
+  }
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_nodelet.cpp:545]: As the inside of for-loop is executed only when ```debug_publisher_ = true```, let's simplify the if statement branch.
```suggestion
  if (debug_publisher_) {
    const double cyclic_time_ms = stop_watch_ptr_->toc("cyclic_time", true);
    const double processing_time_ms = stop_watch_ptr_->toc("processing_time", true);
    debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
      "debug/cyclic_time_ms", cyclic_time_ms);
    debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
      "debug/processing_time_ms", processing_time_ms);

    for (const auto & e : cloud_stdmap_) {
      if (e.second != nullptr) {
        const auto pipeline_latency_ms =
          std::chrono::duration<double, std::milli>(
            std::chrono::nanoseconds(
              (this->get_clock()->now() - e.second->header.stamp).nanoseconds()))
            .count();
        debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
          "debug" + e.first + "/pipeline_latency_ms", pipeline_latency_ms);
      }
    }
  }
```

File: sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp [hunk 10]
```diff
@@ -302,6 +314,51 @@ void PointCloudConcatenationComponent::publish()
...
+        const auto pipeline_latency_ms =
+          std::chrono::duration<double, std::milli>(
+            std::chrono::nanoseconds(
+              (this->get_clock()->now() - e.second->header.stamp).nanoseconds()))
+            .count();
+        debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+          "debug" + e.first + "/pipeline_latency_ms", pipeline_latency_ms);
+      }
+    }
+  }
```
Comment (sasakisasaki) [inline sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_pointclouds.cpp:359]: Same proposal as [this](https://github.com/autowarefoundation/autoware.universe/pull/8535/files#r1746771145).

**Comment by knzo25** [comment]
I know that this is still in draft mode but I would want to avoid people investing their time in vain if we can help it.
@vividf @drwnz @badai-nguyen 

Concatenation in itself should be for more than one node. If you have only one, you can delete it from the launcher. Using this node with one lidar would waste computational resources and increase latency.
Seeing the code changes, even if we go with this approach of adding single pointcloud compatibility, it should be solved with a better design rather just patching. @vividf if working on a redesign right now so if adopted it could be implemented there.

As @badai-nguyen mentioned, other nodes already need the pointcloud to be in the base frame (notably the undistortion) so there is not need to use this node for that purpose

**Comment by shtokuda** [comment]
Thanks for your comment!
The purpose of this PR was to improve convenience by allowing users to configure a single lidar system without having to rewire nodes.
To avoid conflicts with other changes, we will table this change for now.

**Comment by drwnz** [comment]
Thank you for investigating this.
I agree with @knzo25 and @badai-nguyen, there should be no need to use the concat node with a single LiDAR.
While changing the system from multiple to one LiDAR then does require modification of your launch and node configuration, I believe that this is the correct process. Attempting to support a single LiDAR in a node designed specifically to handle multiple LiDARs could introduce more bugs and maintenance issues.

**Pull request closed without merging**